### ISSUE TITLE
Drush without security limitations

### ISSUE THREAD

**Issue by Jimmy2346** [drush-ops/drush#4080]
Hello everyone,

first off: This is not an issue report. I just need clarification about an issue I've been having with Drush for a long time: I always thought Drush is a certain danger to a Drupal site's or server's security.

When run as root and without PHP security features like `open_basedir` and `disable_functions` (which are not compatible with Drush), you can get into problems quickly. Even if you go to great lengths to secure Drupal installations, running, for example, `drush cron` could become very dangerous: Worst case: Bad code (for example from a compromised module, Drupalgeddon-like exploits etc.) runs as root, without any PHP limitations.

This is why I've always modded Drush for my needs in the past. But with Drush 8 (and 9) this seems to become more difficult or even impossible.

That why I'm posting here: Am I not thinking about something here?

**Comment by weitzman** [comment]
You can run drush as any unix user you'd like. Running as root is not typical, though I'm sure some people do that. Many folks use the web server user (e.g. `www-data`) since that user has write perms to uploaded files.

**Comment by Jimmy2346** [comment]
Yes, but that still allows bad code to basically bash-login to my server - without root access, but still. I get it, it's probably much harder to offer Drush's great features without using exec() and similar functions. But unfortunately for me, I won't use it without being able to apply the same security measures as I do for the webserver. Because it runs the same code, if things go wrong.

**Comment by weitzman** [comment]
We have actually reduced the usage of exec() (now Symfony Process which is built on top of proc_open()) a lot in Drush 9.6. It is only used for the sql:* commands. If can give those up, you can disable all the exec() functions.

**Comment by Jimmy2346** [comment]
That would be awesome, but proc_open would also need to be replaced by something else. I will always use Drush on my dev server, so I will be able to keep an eye on it. As soon as it works with open_basedir restrictions in effect and without the common exec-like functions, I will be installing it on the prod machine again ;)

**Comment by weitzman** [comment]
I'd like to point out that not even an SA05 exploit (aka Drupalgeddon) gets the user root control. They can SQL inject, drop all tables, etc but they 't change your PHP code. Once a web site is running other people's PHP, you are already very badly compromised.
